src/server/renderToStream/createBuffer.ts

```diff
@@ -58,66 +49,56 @@ function createBuffer(
         )}`,
       )
     }
-    buffer.push({ chunk, flush: options?.flush })
-    await flushBuffer()
+    const userChunkPromisePrevious = userChunkPromise
+    userChunkPromise = (async () => {
+      if (userChunkPromisePrevious) await userChunkPromisePrevious
+      if (state !== 'STREAMING')
+        await new Promise<void>((resolve) => {
+          onHead = resolve
+        })
+      if (isPromise(chunk)) chunk = await chunk
+      writeChunk(chunk, options?.flush)
+    })()
   }
 
-  async function flushBuffer() {
-    if (!writePermission) {
-      return
-    }
-    if (buffer.length === 0) {
-      return
-    }
-    if (bufferIsFlushing) {
-      await bufferIsFlushing
-      return
-    }
+  function writeChunk(chunk: unknown, flush?: boolean) {
     if (state !== 'STREAMING') {
       assert(state === 'UNSTARTED')
       return
     }
-    let flushStream = false
-    let resolve: () => void
-    bufferIsFlushing = new Promise((r) => (resolve = r))
-    while (buffer.length > 0) {
-      let { chunk, flush } = buffer.shift()!
-      assert(streamOperations.operations)
-      const { writeChunk } = streamOperations.operations
-      if (isPromise(chunk)) chunk = await chunk
-      writeChunk(chunk)
-      if (flush) flushStream = true
-    }
-    bufferIsFlushing = false
-    resolve!()
-    assert(buffer.length === 0)
     assert(streamOperations.operations)
-    if (flushStream && streamOperations.operations.flush !== null) {
+    streamOperations.operations.writeChunk(chunk)
+    if (flush && streamOperations.operations.flush !== null) {
       streamOperations.operations.flush()
       debug('stream flushed')
     }
   }
 
   async function onReactWrite(chunk: unknown) {
-    state === 'UNSTARTED' && debug('>>> START')
+    const firstWrite = state === 'UNSTARTED'
+    firstWrite && debug('>>> START')
     if (debug.isEnabled) {
       debug('react write', getChunkAsString(chunk))
     }
-    state = 'STREAMING'
-    const bufferReactEntry = { chunk: chunk as any, flush: true }
-    if (!writePermission) {
-      buffer.unshift(bufferReactEntry)
+
+    // See Rule 2: https://github.com/brillout/react-streaming/tree/main/src#rule-2
+    if (firstWrite) {
+      state = 'STREAMING'
+      onHead()
     } else {
-      buffer.push(bufferReactEntry)
+      await userChunkPromise
```

Comment:
I think this is not an issue, because `onReactWrite` is called in sync.

Example:
```ts
    for (const reactChunk of reactChunks) {
      onReactWrite(reactChunk)
    }
    // At this point all of the onReactWrite calls are stuck in await userChunkPromise
    // all of are awaiting for the same promise
    injectToStream(userChunk) // write scheduled after all onReactWrite resolves
    injectToStream(userChunk) // write scheduled after all onReactWrite resolves
    injectToStream(userChunk) // write scheduled after all onReactWrite resolves
    for (const reactChunk of reactChunks) {
      onReactWrite(reactChunk)
    }
    // At this point all of the onReactWrite calls are stuck in await userChunkPromise
    // all of are awaiting for the same promise
    // onReactWrite write scheduled after all await userChunkPromise resolves
    injectToStream(userChunk) // write scheduled after all onReactWrite resolves
    ```
    
    So the order of the writes is the same as the order of function calls (`onReactWrite` and `injectToStream`), except if `injectToStream` is called before the first react chunk(s), then it is swapped.

Reply:
The thing I ain't sure about is this:

```js
let resolve
const promise = new Promise(r => resolve = r)
promise.then(() => { console.log(1) })
promise.then(() => { console.log(2) })
promise.then(() => { console.log(3) })
resolve()
```

Is the log guaranteed to be the following?

```
1
2
3
```

Reply:
Yes, I think this is guaranteed.

Reply:
Ok good to know.

Anyways, I refactored that and it's now explicitly in order.